query/workmanager.go

```diff
@@ -321,11 +323,50 @@ Loop:
 				}
 
 			// If the query ended with any other error, put it back
-			// into the work queue.
+			// into the work queue if it has not reached the
+			// maximum number of retries.
 			case result.err != nil:
 				// Punish the peer for the failed query.
 				w.cfg.Ranking.Punish(result.peer.Addr())
 
+				if batch != nil && !batch.noRetryMax {
+					result.job.tries++
+				}
+
+				// Check if this query has reached its maximum
+				// number of retries. If so, remove it from the
+				// batch and don't reschedule it.
+				if batch != nil && !batch.noRetryMax &&
+					result.job.tries >= batch.maxRetries {
+
+					log.Warnf("Query(%d) from peer %v "+
+						"failed and reached maximum "+
+						"number of retries, not "+
+						"rescheduling: %v",
+						result.job.index,
+						result.peer.Addr(), result.err)
+
+					// Decrement the number of queries
+					// remaining in the batch.
+					batch.rem--
+					log.Tracef("Remaining jobs for batch "+
+						"%v: %v ", batchNum, batch.rem)
+
+					if batch.rem > 0 {
+						break
```

Comment:
`break` in a nested loop/select is always a bit tricky to understand. Wouldn't this have the same effect as `continue Loop` and should therefore be replaced with that to make the behavior more clear?

Reply:
A _named_ break (so using a label), might also be useful here.